### PR TITLE
fix wrong ordering in once

### PR DESCRIPTION
**Pull request by matklad**
I believe `Relaxed` on error path is too weak here, and we need Acquire.
On the `Err` path, we use the `old` value to update `current_state`.
With `Relaxed`, I _think_ this can lead to a situation where we add a
partially-initialised node to the waiters list. The waker thread then
will correctly read *Waiter, but `*Waiter.thread` might be garbage.

With an Acquire here we make sure that the `next` field in

    next: current_state.with_addr(current_state.addr() & !STATE_MASK) as *const Waiter,

points to a fully initialized node, which we can then safely read with

`let next = (*queue).next;`

In other words, we want this `Acquire` to synchronize with `AcqRel` in
the same `compare_exchange` statement (on a different thread).

The Release ordering was introduced in

https://github.com/rust-lang/rust/pull/65719/commits/7f1e166899a90226480d564549c36a395e2d8f47

r? @ghost

### CODE REVIEW

**Comment by matklad** [comment]
r? @Amanieu 

Not that I know what I am doing here, the above might be completely wrong :)

**Comment by matklad** [comment]
cc @pitdicker

**Comment by Amanieu** [comment]
I'm pretty sure the `Acquire` is not needed here. All atomic operations on a single atomic variable create a total order of operations on that variable. This means that, no matter which thread looks at it, pushing our thread onto the list always happens after the previous thread pushed itself onto the list. Combined with the release ordering which ensures that setting up the `Waiter` happens before `state_and_queue` is updated, this ensures that the acquire (`AcqRel` in the code) operation after running the initialization function will see all the waiters properly initialized.

**Comment by talchas** [comment]
In particular, release sequences mean that an Acquire of a later thread synchronizes-with not just the last thread's Release, but every previous thread's Release RMW, so you don't need to have each thread synchronize-with the previous thread (which would require this patch).

**Comment by matklad** [comment]
Thanks! 

I now convinced myself that this is ok as well.

It seems that it really doesn’t matter what we do on the err branch: we might as well just be *guessing* the value of the pointer in the loop. Such pointer would be useless on the guessing thread, but, when exchange succeeds, it’ll work for the thread doing wakeups.
